### ISSUE TITLE
Аn unexpected behaviour when using `<transition>` with a Vue component.

### ISSUE THREAD

**Issue by antixrist** [vuejs/vue#4921]
### Vue.js version
2.1.10

### Reproduction Link
https://jsfiddle.net/bhrztdjn/4/

### Steps to reproduce
Click on "Toggle" button.

### What is Expected?
Transitions for a component should have the same behavior as transitions applied to HTML element.

### What is actually happening?
A CSS transition works fine with an HTML element with both directive - `v-if` and `v-show` (as [declared in docs](https://vuejs.org/v2/guide/transitions.html#Transitioning-Single-Elements-Components)).
But for the component, which has a directive v-show on its root element the transition isn't applied on entering and leaving states.
And for the component, which has a directive v-if on its root element the transition isn't applied on leaving state.

**Comment by posva** [comment]
Thanks for opening the issue. Even though I'd add the V-if/v-show in the parent template (https://jsfiddle.net/posva/bhrztdjn/5/), the behaviour is not consistent and at the very least docs should be updated to add a warning about it.

Why do you add the `visible` prop, are you using it inside of the component for other things? Do you mind sharing a bit more about the component itself?

**Comment by antixrist** [comment]
> Why do you add the `visible` prop, are you using it inside of the component for other things?

Yes. For example It needed in modal component. Watcher on the `visible` prop allow to do some internal things like add/remove event listeners to `body` (for `Esc` behavior), add/remove some html classes to <html> and etc. It difficult without `visible` prop.

**Comment by Plasmatium** [comment]
@posva , is this the expected behavior? Is the tag allowed transition only if follow `<transition>` tag 
@antixrist , I think maybe we can pass the transition type intention into template.

```
template: `
<transition :name='transtype'>
<div v-show="visible">I am a component with v-show directive (inside transition)</div>
</transition>
`,
props: [ 'visible', 'transtype']
```

Here is the code: [jsfiddle](https://jsfiddle.net/Plasmatium/bhrztdjn/13/)

**Comment by gebilaoxiong** [comment]
@antixrist 

Dear antixrist  : )

I think this is not a good idea

because `transition` is a component,  `cmp-v-if`(child component) is also a component

it looks like you want to use `child component` state to control the parent

**Comment by antixrist** [comment]
@gebilaoxiong nope.
Child component can take `value` prop for control it from parent by `v-model` and can emit `input` event.
So only parent component decides about change state and the children is aware about this state.

[Here is](https://gist.github.com/antixrist/4458c76b34c183c3fa93975d39bdd12b) my modal component with example.
In this component data flow is not broken, but I still can not manage transitions outside of the `<modal>` component. Although the documentation [says otherwise](https://vuejs.org/v2/guide/transitions.html#Transitioning-Single-Elements-Components) (which is confirmed by a @posva):

> Vue provides a transition wrapper component, allowing you to add entering/leaving transitions for any element or component in the following contexts:
> * Conditional rendering (using v-if)
> * Conditional display (using v-show)
> * Dynamic components
> * Component root nodes

**Comment by Nartub600** [comment]
I had a transition which worked with v-if but not with v-show. I had to resort to this solution https://medium.com/vuejs-tips/css-only-v-show-fade-animation-6f7818fdff4 for it to work

**Comment by YavuzSecer** [comment]
It appears that this issue is back. Im having the same warningson version 2.6.6

**Comment by WofWca** [comment]
I think "transition behavior is inconsistent for v-show/v-if and components/elements" would be a better title. Would make it easier to find it.

**Comment by RajithaKumara** [comment]
Hi @antixrist 
Did you managed the modal component? If so, I want some help
